crownheightsaid/mutual-aid-app
# Add /delivery-needed API

## Summary
- Add new endpoint for fetching all requests from Airtable that is marked `Delivery Needed` and has been posted to Slack.
- Add lru-cache to cache geo lookups.

I decided to use in-memory LRU caching as a quick solution rather than caching the lookups in Airtable. The reasons being:
a) I don't want to be making writes to Airtable as it might take awhile if there are many requests.
b) Will need to handle checking when the cross street data has been updated.
c) Cross streets data overlaps. Plus this geo data can be reused in other APIs if we store it this way.

I also decided to return data in GeoJSON format, storing any peripheral data in `properties.meta`. This will make it much easier to display on a map with clustering, plus since it's an open standard and compatible with many mapping tools, this API can be reused elsewhere.

Lookups are  set to cache for 7 days. 

### Manual testing
Start the server and visit http://localhost:5000/api/delivery-needed/requests.json.
On first visit, request will take longer ~300ms for about 5 requests, but will be much quicker ~100ms on subsequent visits.

Example payload:
```
{
  "type": "FeatureCollection",
  "features": [
    {
      "type": "Feature",
      "geometry": {
        "type": "Point",
        "coordinates": [
          -73.94984500000001,
          40.6768827
        ]
      },
      "properties": {
        "title": "RRGLAK",
        "meta": {
          "Code": "RRGLAK",
          "Cross Street #1": "Dean St",
          "Cross Street #2": "Nostrand Ave",
          "Neighborhood Area (See Map)": "NW"
        }
      }
    },
    {
      "type": "Feature",
      "geometry": {
        "type": "Point",
        "coordinates": [
          -73.9308727,
          40.6714765
        ]
      },
      "properties": {
        "title": "HOJQJK",
        "meta": {
          "Code": "HOJQJK",
          "Cross Street #1": "Utica Ave",
          "Cross Street #2": "Sterling St",
          "Neighborhood Area (See Map)": "NE"
        }
      }
    }
}
```

### Performance
timing data on first hit:
<img width="214" alt="Screen Shot 2020-04-26 at 11 49 20 AM" src="https://user-images.githubusercontent.com/1868400/80312614-0cfd2080-87b4-11ea-81a3-80400137ff8b.png">

subsequent hits:
<img width="233" alt="Screen Shot 2020-04-26 at 11 49 26 AM" src="https://user-images.githubusercontent.com/1868400/80312623-11293e00-87b4-11ea-9bb9-7bd5c5edab31.png">
